storage/src/journal/fixed.rs

```diff
@@ -0,0 +1,656 @@
+//! consisting entirely of values older than the given item.
+//!
+//! # Replay
+//!
+//! The `replay` method iterates over multiple blobs concurrently to support fast reading of all
+//! unpruned items into memory.
+
+use super::Error;
+use bytes::BufMut;
+use commonware_runtime::{Blob, Error as RError, Storage};
+use commonware_utils::hex;
+use futures::stream::{self, Stream, StreamExt};
+use prometheus_client::metrics::{counter::Counter, gauge::Gauge};
+use prometheus_client::registry::Registry;
+use std::sync::{Arc, Mutex};
+use tracing::{debug, trace, warn};
+
+/// Configuration for `Journal` storage.
+#[derive(Clone)]
+pub struct Config {
+    /// Registry for metrics.
+    pub registry: Arc<Mutex<Registry>>,
+
+    /// The `commonware-runtime::Storage` partition to use for storing journal blobs.
+    pub partition: String,
+
+    pub items_per_blob: u64,
+}
+
+/// Implementation of `Journal` storage.
+pub struct Journal<B: Blob, E: Storage<B>, const N: usize> {
+    runtime: E,
+    cfg: Config,
+
+    oldest_unpruned_blob: usize,
+    blobs: Vec<B>,
+
+    tracked: Gauge,
+    synced: Counter,
+    pruned: Counter,
+}
+
+impl<B: Blob, E: Storage<B>, const N: usize> Journal<B, E, N> {
+    /// Initialize a new `Journal` instance.
+    ///
+    /// All backing blobs are opened but not read during initialization. The `replay` method can be
+    /// used to iterate over all items in the `Journal`.
+    pub async fn init(runtime: E, cfg: Config) -> Result<Self, Error> {
+        // Iterate over blobs in partition
+        let mut blobs: Vec<(B, u64)> = Vec::new();
+        let stored_blobs = match runtime.scan(&cfg.partition).await {
+            Ok(blobs) => blobs,
+            Err(RError::PartitionMissing(_)) => Vec::new(),
+            Err(err) => return Err(Error::Runtime(err)),
+        };
+        for name in stored_blobs {
+            let blob = runtime
+                .open(&cfg.partition, &name)
+                .await
+                .map_err(Error::Runtime)?;
+            let index = match name.try_into() {
+                Ok(index) => u64::from_be_bytes(index),
+                Err(nm) => return Err(Error::InvalidBlobName(hex(&nm))),
+            };
+            debug!(blob = index, "loaded blob");
+            blobs.push((blob, index));
+        }
+        blobs.sort_by(|(_, a), (_, b)| a.cmp(b));
+        let mut oldest_blob = 0;
+        if !blobs.is_empty() {
+            // Check that there are no gaps in the blob numbering, which would indicate missing data.
+            let mut it = blobs.iter();
+            let mut previous_blob = it.next().unwrap().1;
+            oldest_blob = previous_blob;
+            for (_, section) in it {
+                if *section != previous_blob + 1 {
+                    return Err(Error::MissingBlob(previous_blob + 1));
+                }
+                previous_blob = *section;
+            }
+        } else {
+            debug!("no blobs found, bootstrapping with 1 empty blob");
```

Comment:
nit: I would just leave it as `no blobs found`

Reply:
done